lib/MongoDB/_Protocol.pm

```diff
@@ -86,8 +88,150 @@ use constant {
     # length for OP_COMPRESSED
     P_COMPRESSED_PREFIX_LENGTH =>
         length(pack P_COMPRESSED, 0, 0, 0, 0, 0, 0, 0),
+    P_MSG_PREFIX_LENGTH =>
+        length(pack P_MSG, 0, 0, 0, 0, 0),
 };
 
+# struct OP_MSG {
+#     MsgHeader header;             // standard message header, with opCode 2013
+#     uint32    flagBits;
+#     Section+  sections;
+#     [uint32   checksum;]
+# };
+#
+# struct Section {
+#     uint8 payloadType;
+#     union payload {
+#         document document;        // payloadType == 0
+#         struct   sequence {       // payloadType == 1
+#             int32     size;
+#             cstring   identifier;
+#             document* documents;
+#         };
+#     };
+# };
+
+use constant {
+    P_SECTION_PAYLOAD_TYPE  => "C",
+    P_SECTION_SEQUENCE_SIZE => PERL58 ? "l" : "l<",
+};
+
+use constant {
+    P_SECTION_HEADER               => P_SECTION_PAYLOAD_TYPE . P_SECTION_SEQUENCE_SIZE,
+    P_SECTION_PAYLOAD_TYPE_LENGTH  => length( pack P_SECTION_PAYLOAD_TYPE, 0 ),
+    P_SECTION_SEQUENCE_SIZE_LENGTH => length( pack P_SECTION_SEQUENCE_SIZE, 0 ),
+};
+
+use MongoDB::Protocol::_Section;
+
+=method prepare_sections( $cmd )
+
+Takes a command, returns sections ready for joining
+
+=cut
+
+sub prepare_sections {
```

Comment:
Instead of washing this through objects, I think you can just directly return 1 or more encoded strings.  The protocol 0 part is a type byte concatenated with either an encoded document verbatim, or an encoded document with a key removed.  The protocol 1 part is either missing or is type and length bytes concatenated with all the encoded documents of the extracted field.

In hash terms (not that you can use hashes for this) I think it's along these lines:
```
my $copy = { %$cmd };
my $extracted = delete $copy{$special_key};
my @sections = pack(..., 0, $codec->encode_one($copy)); # payload type 0
if ( $extracted ) {
    my $payload1 = pack(..., 1, 0, $special_key, join('', map { $codec->encode_one($_) } @$extracted)); 
    # fix up bytes 2-5 with payload length
    push @sections, $payload1;
}
```